pkg/operator/tlsrotationcontroller.go

```diff
@@ -0,0 +1,166 @@
+	fmt.Println(">>> reason ", reason)
+	fmt.Printf(">>> "+messageFmt+"\n", args...)
+}
+
+func (e *eventRecorder) Warning(reason, message string) {
+	fmt.Println("*** reason:", reason)
+	fmt.Println("*** message:", message)
+}
+
+func (e *eventRecorder) Warningf(reason, messageFmt string, args ...interface{}) {
+	fmt.Println("*** reason ", reason)
+	fmt.Printf("*** "+messageFmt+"\n", args...)
+}
+
+func (e *eventRecorder) Shutdown() {}
+
+// ForComponent allows to fiddle the component name before sending the event to sink.
+// Making more unique components will prevent the spam filter in upstream event sink from dropping
+// events.
+func (e *eventRecorder) ForComponent(componentName string) events.Recorder {
+	newEr := eventRecorder(componentName)
+	return &newEr
+}
+
+// WithComponentSuffix is similar to ForComponent except it just suffix the current component name instead of overriding.
+func (e *eventRecorder) WithComponentSuffix(componentNameSuffix string) events.Recorder {
+	newEr := eventRecorder(string(*e) + "-" + componentNameSuffix)
+	return &newEr
+}
+
+// ComponentName returns the current source component name for the event.
+func (e *eventRecorder) ComponentName() string {
+	return string(*e)
+}
+
+type TLSRotationController struct {
+	name, namespace string
+	hostnames       []string
+	kubeClient      *kubernetes.Clientset
+	informer        informers.SharedInformerFactory
+	eventRecorder   eventRecorder
+	controller      factory.Controller
+	ctx             context.Context
+}
+
+func NewTLSRotationController(config *rest.Config,
+	namespace string,
+	name string,
+	hostnames []string) (*TLSRotationController, error) {
+
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	inf := informers.NewSharedInformerFactoryWithOptions(
+		kubeClient,
+		1*time.Hour,
+		informers.WithNamespace(namespace),
+	)
+
+	er := eventRecorder("cmo" + name)
+
+	crc := certrotation.NewCertRotationController(
+		name+"CertRotationController",
+		certrotation.SigningRotation{
+			Namespace:     namespace,
+			Name:          name + "-signer-ca",
+			Refresh:       15 * 24 * time.Hour,
+			Validity:      30 * 24 * time.Hour,
+			Informer:      inf.Core().V1().Secrets(),
+			Lister:        inf.Core().V1().Secrets().Lister(),
+			Client:        kubeClient.CoreV1(),
+			EventRecorder: &er,
+		},
+		certrotation.CABundleRotation{
+			Namespace:     namespace,
+			Name:          name + "-ca-bundle",
+			Informer:      inf.Core().V1().ConfigMaps(),
+			Lister:        inf.Core().V1().ConfigMaps().Lister(),
+			Client:        kubeClient.CoreV1(),
+			EventRecorder: &er,
+		},
+		certrotation.TargetRotation{
+			Namespace: namespace,
+			Name:      name + "-cert",
+			Refresh:   24 * time.Hour,
+			Validity:  2 * 24 * time.Hour,
+			CertCreator: &certrotation.ServingRotation{
+				Hostnames: func() []string {
+					klog.V(4).Info("get Hostnames")
+					return []string{"prometheus-grpc", "thanos-querier"}
```

Comment:
that should be returning `hostnames`?

Reply:
Currently we are not issuing certificates bound to host names, I will try to persuade the cert controller (or the underlying logic) to do the same.

Reply:
small correction: the underlying machinery requires hostnames to be set, but:

1. the first entry is being interpreted as the common name (CN)
2. the whole list is being expanded as subject alternate DNS names and (and only if resolvable via DNS) the underlying IP addresses.

While for point 2. the subject alternate DNS names are being set, thanos does not verify them, hence we can treat that list logically as a one-element array which includes the common name in its first entry.